// node-id3-reader/lib/frame-identifiers.js
'use strict'

const CODES = [
    'AENC',
    'APIC',
    'ASPI',
    'COMM',
    'COMR',
    'ENCR',
    'EQU2',
    'ETCO',
    'GEOB',
    'GRID',
    'LINK',
    'MCDI',
    'MLLT',
    'OWNE',
    'PRIV',
    'PCNT',
    'POPM',
    'POSS',
    'RBUF',
    'RVA2',
    'RVRB',
    'SEEK',
    'SIGN',
    'SYLT',
    'SYTC',
    'TALB',
    'TBPM',
    'TCOM',
    'TCON',
    'TCOP',
    'TDEN',
    'TDLY',
    'TDOR',
    'TDRC',
    'TDRL',
    'TDTG',
    'TENC',
    'TEXT',
    'TFLT',
    'TIPL',
    'TIT1',
    'TIT2',
    'TIT3',
    'TKEY',
    'TLAN',
    'TLEN',
    'TMCL',
    'TMED',
    'TMOO',
    'TOAL',
    'TOFN',
    'TOLY',
    'TOPE',
    'TOWN',
    'TPE1',
    'TPE2',
    'TPE3',
    'TPE4',
    'TPOS',
    'TPRO',
    'TPUB',
    'TRCK',
    'TRSN',
    'TRSO',
    'TSOA',
    'TSOP',
    'TSOT',
    'TSRC',
    'TSSE',
    'TSST',
    'TYER',
    'TXXX',
    'UFID',
    'USER',
    'USLT',
    'WCOM',
    'WCOP',
    'WOAF',
    'WOAR',
    'WOAS',
    'WORS',
    'WPAY',
    'WPUB',
    'WXXX',
]

module.exports = {
    descriptions : {
        AENC : {
            key : 'encryption',
            text : 'Audio Encryption',
        },
        APIC : {
            key : 'attachedPicture',
            text : 'Attached Picture',
        },
        ASPI : {
            key : 'audioSeekPointIndex',
            text : 'Audio Seek point index',
        },
        COMM : {
            key : 'comments',
            text : 'Comments',
        },
        COMR : {
            key : 'commercialFrame',
            text : 'Commercial frame',
        },
        ENCR : {
            key : 'encryptionMethod',
            text : 'Encryption method registration',
        },
        EQU2 : {
            key : 'equalisation',
            text : 'Equalisation',
        },
        ETCO : {
            key : 'eventTimingCodes',
            text : 'Event timing codes',
        },
        GEOB : {
            key : 'encapObject',
            text : 'General Encapsulated object',
        },
        GRID : {
            key : 'groupIdentification',
            text : 'Group Identification registration',
        },
        LINK : {
            key : 'linked',
            text : 'Linked information',
        },
        MCDI : 'Music CD identifier',
        MLLT : 'MPEG Location lookup table',
        OWNE : 'Ownership frame',
        PRIV : 'Private frame',
        PCNT : 'Play counter',
        POPM : 'Popularimeter',
        POSS : 'Position Synchronisation frame',
        RBUF : 'Recommended Buffer size',
        RVA2 : 'Relative volume adjustment',
        RVRB : 'Reverb',
        SEEK : 'Seek frame',
        SIGN : 'Signature frame',
        SYLT : 'Synchronised lyric/text',
        SYTC : 'Synchronised tempo codes',
        TALB : {
            key : 'album',
            text : 'Album/Movie/Show title',
        },
        TBPM : 'BPM',
        TCOM : 'Composer',
        TCON : {
            key : 'genre',
            text : 'Content Type',
        },
        TCOP : 'Copyright Message',
        TDEN : 'Encoding time',
        TDLY : 'Playlist delay',
        TDOR : 'Original Release time',
        TDRC : 'Recording time',
        TDRL : 'Release time',
        TDTG : 'Tagging time',
        TENC : 'Encoded by',
        TEXT : 'Lyricist/Text writer',
        TFLT : 'File type',
        TIPL : 'Involved people list',
        TIT1 : 'Content group description',
        TIT2 : {
            key : 'title',
            text : 'Title/songname/content description',
        },
        TIT3 : {
            key : 'subtitle',
            text : 'Subtitle/Description refinement',
        },
        TKEY : 'Initial key',
        TLAN : {
            key : 'language',
            text : 'Language(s)',
        },
        TLEN : 'Length',
        TMCL : 'Musician credits list',
        TMED : 'Media type',
        TMOO : 'Mood',
        TOAL : 'Original album/movie/show title',
        TOFN : 'Original filename',
        TOLY : 'Original lyricist/text writer',
        TOPE : 'Original artist/performer',
        TOWN : 'File owner/licensee',
        TPE1 : {
            key : 'artist',
            text : 'Lead performer/Solist',
        },
        TPE2 : {
            key : 'band',
            text : 'Band/Orchestra/accompaniment',
        },
        TPE3 : 'Conductor/performer refinement',
        TPE4 : 'Interpreted, remixed or otherwise modified by',
        TPOS : 'Part of a set',
        TPRO : 'Produced notice',
        TPUB : 'Publisher',
        TRCK : 'Track number/Position in set',
        TRSN : 'Internet Radio Station Name',
        TRSO : 'Internet Radio Station Owner',
        TSOA : 'Album sort order',
        TSOP : 'Performer sort order',
        TSOT : 'Title sort order',
        TSRC : 'International Standard Recording Code',
        TSSE : 'Software/Hardware settings used for encoding',
        TSST : 'Set subtitle',
        TYER : {
            key : 'year',
            text : 'Year',
        },
        TXXX : 'User defined',
        UFID : {
            key : 'ufid',
            text : 'Unique file identifier',
        },
        USER : 'Terms of use',
        USLT : 'Unsynchronised lyric/text transcription',
        WCOM : 'Commercial information',
        WCOP : 'Copyright/Legal information',
        WOAF : 'Official audio file webpage',
        WOAR : 'Official artist/performer webpage',
        WOAS : 'Official audio source webpage',
        WORS : 'Official Internet radio station webpage',
        WPAY : 'Payment',
        WPUB : 'Publishers official webpage',
        WXXX : 'User defined URL link frame',
    },
    CODES,
}
